dialogue-core/src/main/java/com/palantir/dialogue/core/PinUntilErrorChannel.java

```diff
@@ -0,0 +1,260 @@
+        Preconditions.checkArgument(!channels.isEmpty(), "List of channels must not be empty");
+        ConstantNodeList constantNodeList = new ConstantNodeList(channels, ThreadLocalRandom.current());
+        return new PinUntilErrorChannel(constantNodeList);
+    }
+
+    @Override
+    public Optional<ListenableFuture<Response>> maybeExecute(Endpoint endpoint, Request request) {
+        int currentIndex = currentHost.get();
+        LimitedChannel channel = maybeShufflingNodeList.get(currentIndex);
+
+        Optional<ListenableFuture<Response>> maybeFuture = channel.maybeExecute(endpoint, request);
+        if (!maybeFuture.isPresent()) {
+            OptionalInt next = incrementCurrentHost(currentIndex);
+            debugLogCurrentChannelRejected(currentIndex, channel, next);
+            return Optional.empty(); // if the caller retries immediately, we'll get the next host
+        }
+
+        ListenableFuture<Response> future = maybeFuture.get();
+        return Optional.of(DialogueFutures.addDirectCallback(future, new FutureCallback<Response>() {
+            @Override
+            public void onSuccess(Response response) {
+                if (response.code() >= 300) {
+                    OptionalInt next = incrementCurrentHost(currentIndex);
+                    debugLogReceivedErrorStatus(currentIndex, channel, response, next);
+                    // TODO(dfox): handle 308 See Other somehow, as we currently don't have a host -> channel mapping
+                }
+            }
+
+            @Override
+            public void onFailure(Throwable throwable) {
+                OptionalInt next = incrementCurrentHost(currentIndex);
+                debugLogReceivedThrowable(currentIndex, channel, throwable, next);
+            }
+        }));
+    }
+
+    /**
+     * If we have some reason to think the currentIndex is bad, we want to move to the next host. This is done with a
+     * compareAndSet to ensure that out of order responses which signal information about a previous host don't kick
+     * us off a good one.
+     */
+    private OptionalInt incrementCurrentHost(int currentIndex) {
+        int nextIndex = Math.max(currentIndex + 1, 0); // we want Integer.MAX_VALUE to wrap around to zero
+
+        boolean saved = currentHost.compareAndSet(currentIndex, nextIndex);
+        return saved ? OptionalInt.of(nextIndex) : OptionalInt.empty(); // we've moved on already
+    }
+
+    interface NodeList {
+        /**
+         * Accepts positive indexes that are greater than the length of the list, returns an item modulo the length
+         * of the list.
+         * @return
+         */
+        LimitedChannel get(int index);
+    }
+
+    @VisibleForTesting
+    static final class ConstantNodeList implements NodeList {
+        private final List<LimitedChannel> channels;
+
+        ConstantNodeList(List<LimitedChannel> channels, Random random) {
+            this.channels = shuffleImmutableList(channels, random);
```

Comment:
why do we need to shuffle here?

Reply:
This is actually pretty important - if there are 100 independent servers all talking to a few catalog nodes and they all start off with the nodelist in the same order, then when catalog node1 restarts, all that nodes traffic will shift onto node2 and stay there. Node2 will then have double traffic, and when it restarts, they'll all shift to node3.

Making sure that the order is shuffled at startup time means that when a single upstream goes down, the traffic should be distributed across upstreams.

I can add a comment to this effect if you like?

Reply:
ok that totally makes sense. Ya a comment would be helpful thanks!